### PR TITLE
 [chore] add test helper to verify exporter behavior on errors

### PR DESCRIPTION
**Pull request by omrozowicz-splunk**
**Description:** Adding a test helper to test exporter behaviour on errors.

**Link to tracking Issue:** https://github.com/open-telemetry/opentelemetry-collector/issues/7479
Will also fix: https://github.com/open-telemetry/opentelemetry-collector/issues/7481

It is created based on the exporter receiver test: https://github.com/open-telemetry/opentelemetry-collector/pull/7516
The exporter targets `mockReceiver` as an endpoint.

The test cases covered here are:
1. ConsumeLogs/Traces/Metrics call succeeds at the first try
2. ConsumeLogs/Traces/Metrics call fails once with non-permanent error, then succeeds on one of the next tries
3. ConsumeLogs/Traces/Metrics call fails once with permanent error and the data is not retried
4. ConsumeLogs/Traces/Metrics call fails once with permanent error or permanent error (then the test pass requirement is calculated based on number of errors per kind/successfully delivered data/total number of data to be delivered)

### CODE REVIEW

File: exporter/exportertest/contract_checker.go [hunk 1]
```diff
@@ -0,0 +1,298 @@
...
+type CheckConsumeContractParams struct {
+	T *testing.T
+	// Factory that allows to create an exporter.
+	Factory exporter.Factory
+	// DataType to test for.
+	DataType component.DataType
+	// Config of the exporter to use.
+	Config               component.Config
+	NumberOfTestElements int
+	MockReceiver         interface{}
```
Comment (dmitryax) [inline exporter/exportertest/contract_checker.go:39]: This should be an interface providing the functionality needed for the contract checker only. `mockMetricsReceiver` and others should implement this interface in `exporter/otlpexporter` module. 

Probably better to do it thorough a factory
```suggestion
	MockReceiverFactory MockReceiverFactory
```

where 
```golang
type DecisionFunc func() error

type MockReceiverFactory func(DecisionFunc) MockReceiver

type MockReceiver interface {
     ... methods needed for contract checked
}

The main idea is that otlpexporter (or any other exporter) should depend on the contract checker, not the other way.

Comment (omrozowicz-splunk) [reply exporter/exportertest/contract_checker.go]: Hey, I changed it `MockReceiver MockReceiver` for now, I think I'm missing the point of why and how to use `type MockReceiverFactory func(DecisionFunc) MockReceiver`. I keep the decision functions in contract checker to mimic the same behaviour that was implemented in receiver tests: https://github.com/open-telemetry/opentelemetry-collector/pull/7516/files#diff-a246dccd935c825bf517eb70b6476fdeb05f4750d960247b7df7fbf9a17b23aeR85
I would appreciate some more tips on how to use `MockReceiverFactory` ☺️

Comment (dmitryax) [reply exporter/exportertest/contract_checker.go]: The main goal is to move the actual test using the helper to the tested modules, e.g. `TestConsumeContractOtlpLogs` should go to `exporter/otlpexporter` module. Keeping that in mind should influence the right decisions for the `MockReceiver` interface. 

You need to pass DecisionFunc to the exporter test which should give you an implementation of `MockReceiver` in return. `MockReceiverFactory` should be implemented on the exporter side. Please let me know if it makes it any clearer.

Comment (omrozowicz-splunk) [reply exporter/exportertest/contract_checker.go]: Ok, I think now it is more clear, I pushed another commit with MockReceiverFactory and tests moved under `exporter/otlpexproter`. It took some gymnastics with `net.Listen` (as now we create an instance every time depending on decision func). Please take a look and tell me if is it more or less what you meant.

Comment (dmitryax) [reply exporter/exportertest/contract_checker.go]: We are going in the right direction. Half way through :)

File: exporter/exportertest/contract_checker.go [hunk 2]
```diff
@@ -0,0 +1,269 @@
...
+
+	for i := 0; i < params.NumberOfTestElements; i++ {
+		id := UniqueIDAttrVal(strconv.Itoa(i))
+		fmt.Println("Preparing metric number: ", id)
+		data := CreateOneMetricWithID(id)
+
+		err = exp.ConsumeMetrics(ctx, data)
+	}
+
+	reqCounter := mockReceiver.getReqCounter()
```
[the remark this answers is not in the document]
Comment (omrozowicz-splunk) [reply exporter/exportertest/contract_checker.go:127]: For now I set up retrySettings to 10 miliseconds with these settings:

```
// NewTestRetrySettings returns the default settings for otlp exporter test.
func NewTestRetrySettings() exporterhelper.RetrySettings {
	return exporterhelper.RetrySettings{
		Enabled: true,
		// interval is short for the test purposes
		InitialInterval:     10 * time.Millisecond,
		RandomizationFactor: backoff.DefaultRandomizationFactor,
		Multiplier:          1.1,
		MaxInterval:         10 * time.Second,
		MaxElapsedTime:      1 * time.Minute,
	}
}
```

but yes, I can add delay, we might have problems if we're very unlucky and getting error all the time 😅

File: exporter/exportertest/mock_receiver.go [hunk 3]
```diff
@@ -0,0 +1,355 @@
...
+var errPermanent = status.Error(codes.Internal, "Permanent error")
+
+type DecisionFunc func() error
+
+type MockReceiver interface {
+	clearCounters()
+	setExportErrorFunction(decisionFunction func() error)
+	srvStop()
+	getListenerAddress() string
+	getReqCounter() requestCounter
```
[the remark this answers is not in the document]
Comment (omrozowicz-splunk) [reply exporter/exportertest/mock_receiver.go]: ok, thanks for the tips! so I deleted Exports and put `ConsumeX` instead, and created a `mockConsumer`. I reused some of the code from the link. Additionally I keep there `ReceivedTraces/Metrics/Logs`, but not sure if I can make it useful when I already have that requestCounter - most of the logic is based on counting how many errors of a kind do we have. I can use the len of `ReceivedX`, but still I need to keep number of requests ent/failed somewhere.

File: connector/forwardconnector/go.mod [hunk 4]
```diff
@@ -76,3 +76,19 @@ retract (
...
+
+replace go.opentelemetry.io/collector/config/internal => ../../config/internal
+
+replace go.opentelemetry.io/collector/extension/auth => ../../extension/auth
+
+replace go.opentelemetry.io/collector/exporter/otlpexporter => ../../exporter/otlpexporter
+
+replace go.opentelemetry.io/collector/config/configopaque => ../../config/configopaque
+
+replace go.opentelemetry.io/collector/config/configauth => ../../config/configauth
```
[the remark this answers is not in the document]
Comment (omrozowicz-splunk) [reply connector/forwardconnector/go.mod]: Right, it disappeared 😄

File: exporter/exportertest/mock_receiver.go [hunk 5]
```diff
@@ -0,0 +1,355 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+package exportertest // import "go.opentelemetry.io/collector/exporter/exportertest"
+
+import (
+	"context"
+	"fmt"
+	"go.opentelemetry.io/collector/component"
+	"go.opentelemetry.io/collector/exporter"
```
[the remark this answers is not in the document]
Comment (omrozowicz-splunk) [reply exporter/exportertest/mock_receiver.go]: done

File: exporter/exportertest/mock_receiver.go [hunk 6]
```diff
@@ -0,0 +1,355 @@
...
+// // half of the time and fails with a non-permanent error the rest of the time.
+func randomNonPermanentErrorConsumeDecision() error {
+	if rand.Float32() < 0.5 {
+		return errNonPermanent
+	}
+	return nil
+}
+
+// Define a function that matches the MockReceiverFactory signature
+func CreateMockOtlpReceiver(decisionFunc DecisionFunc, dataType component.DataType, ln net.Listener) MockReceiver {
```
[the remark this answers is not in the document]
Comment (omrozowicz-splunk) [reply exporter/exportertest/mock_receiver.go]: done

File: exporter/otlpexporter/consume_contract_test.go [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+	"go.opentelemetry.io/collector/config/configgrpc"
+	"go.opentelemetry.io/collector/config/configtls"
+	"go.opentelemetry.io/collector/exporter/exporterhelper"
+	"go.opentelemetry.io/collector/exporter/exportertest"
+)
+
+const DefaultOTLPPort = 9999
+
+// NewTestRetrySettings returns the default settings for otlp exporter test.
+func NewTestRetrySettings() exporterhelper.RetrySettings {
```
Comment (dmitryax) [inline exporter/otlpexporter/consume_contract_test.go:25]: There is no need to export this function

File: exporter/otlpexporter/consume_contract_test.go [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+	"github.com/cenkalti/backoff/v4"
+
+	"go.opentelemetry.io/collector/component"
+	"go.opentelemetry.io/collector/config/configgrpc"
+	"go.opentelemetry.io/collector/config/configtls"
+	"go.opentelemetry.io/collector/exporter/exporterhelper"
+	"go.opentelemetry.io/collector/exporter/exportertest"
+)
+
+const DefaultOTLPPort = 9999
```
Comment (dmitryax) [inline exporter/otlpexporter/consume_contract_test.go:22]: This one should not be exported as well

File: exporter/otlpexporter/consume_contract_test.go [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+		// interval is short for the test purposes
+		InitialInterval:     10 * time.Millisecond,
+		RandomizationFactor: backoff.DefaultRandomizationFactor,
+		Multiplier:          1.1,
+		MaxInterval:         10 * time.Second,
+		MaxElapsedTime:      1 * time.Minute,
+	}
+}
+
+func ReturnOtlpConfig(port int) component.Config {
```
Comment (dmitryax) [inline exporter/otlpexporter/consume_contract_test.go:37]:
```suggestion
func testConfig(port int) component.Config {
```

File: exporter/otlpexporter/consume_contract_test.go (hunk 8 above)
Comment (dmitryax) [reply exporter/otlpexporter/consume_contract_test.go:22]: btw why 9999? We can extract the real default port from `defaultGRPCEndpoint`

File: exporter/otlpexporter/mock_receiver.go [hunk 10]
```diff
@@ -0,0 +1,78 @@
...
+
+type BaseOTLPDataReceiver struct {
+	DataReceiverBase
+	mockConsumer    *exportertest.MockConsumer
+	traceReceiver   receiver.Traces
+	metricsReceiver receiver.Metrics
+	logReceiver     receiver.Logs
+}
+
+func NewOTLPDataReceiver(port int, mockConsumer *exportertest.MockConsumer) *BaseOTLPDataReceiver {
```
Comment (dmitryax) [inline exporter/otlpexporter/mock_receiver.go:32]: this should not be exported

File: exporter/otlpexporter/mock_receiver.go [hunk 11]
```diff
@@ -0,0 +1,78 @@
...
+	"go.opentelemetry.io/collector/receiver/receivertest"
+)
+
+// DataReceiverBase implement basic functions needed by all receivers.
+type DataReceiverBase struct {
+	// Port on which to listen.
+	Port int
+}
+
+type BaseOTLPDataReceiver struct {
```
Comment (dmitryax) [inline exporter/otlpexporter/mock_receiver.go:24]: No need to export publicly 
```suggestion
type mockReceiver struct {
```

File: exporter/otlpexporter/mock_receiver.go [hunk 12]
```diff
@@ -0,0 +1,78 @@
...
+	"go.opentelemetry.io/collector/component/componenttest"
+	"go.opentelemetry.io/collector/config/confignet"
+	"go.opentelemetry.io/collector/exporter/exportertest"
+	"go.opentelemetry.io/collector/receiver"
+	"go.opentelemetry.io/collector/receiver/otlpreceiver"
+	"go.opentelemetry.io/collector/receiver/receivertest"
+)
+
+// DataReceiverBase implement basic functions needed by all receivers.
+type DataReceiverBase struct {
```
Comment (dmitryax) [inline exporter/otlpexporter/mock_receiver.go:19]: Why is this needed as a separate struct?

File: exporter/exportertest/contract_checker.go [hunk 13]
```diff
@@ -0,0 +1,260 @@
...
+
+	reqCounter := mockReceiver.RequestCounter()
+	// The overall number of requests sent by exporter
+	fmt.Printf("Number of export tries: %d\n", reqCounter.total)
+	// Successfully delivered items
+	fmt.Printf("Total items received successfully: %d\n", reqCounter.success)
+	// Number of errors that happened
+	fmt.Printf("Number of permanent errors: %d\n", reqCounter.error.permanent)
+	fmt.Printf("Number of non-permanent errors: %d\n", reqCounter.error.nonpermanent)
+	checkIfTestPassed(params.T, params.NumberOfTestElements, reqCounter)
```
Comment (dmitryax) [inline exporter/exportertest/contract_checker.go:206]: We cannot check this synchronously. The test should pass with the enabled exporter's queue as well. Please use `assert.Eventually` instead

File: exporter/otlpexporter/consume_contract_test.go (hunk 7 above)
Comment (omrozowicz-splunk) [reply exporter/otlpexporter/consume_contract_test.go]: done

File: exporter/otlpexporter/consume_contract_test.go (hunk 8 above)
Comment (omrozowicz-splunk) [reply exporter/otlpexporter/consume_contract_test.go]: I think it was a leftover form testing, now I left default grpc settings in the constructor, but to use `defaultGRPCEndpoint` in the config passed to exporter I think I'd need to make it exportable, right? I hardcoded it for now, I'm 99% sure it's wrong but don't know what would be clean way to do this.

Comment (omrozowicz-splunk) [reply exporter/otlpexporter/consume_contract_test.go]: Or do you mean that I should use it like this?

File: exporter/otlpexporter/consume_contract_test.go (hunk 9 above)
Comment (omrozowicz-splunk) [reply exporter/otlpexporter/consume_contract_test.go]: done

File: exporter/otlpexporter/mock_receiver.go (hunk 10 above)
Comment (omrozowicz-splunk) [reply exporter/otlpexporter/mock_receiver.go]: done

File: exporter/otlpexporter/mock_receiver.go (hunk 11 above)
Comment (omrozowicz-splunk) [reply exporter/otlpexporter/mock_receiver.go]: Done, but `mockReceiver` already existed in this package in `otlp_test`, so I named it `mockOtlpReceiver`

File: exporter/otlpexporter/mock_receiver.go (hunk 12 above)
Comment (omrozowicz-splunk) [reply exporter/otlpexporter/mock_receiver.go]: I used the structure from contrib's testbed, but it indeed does not help here, I deleted this part and port is set up by the default config of otlp receiver factory

File: exporter/exportertest/contract_checker.go (hunk 13 above)
Comment (omrozowicz-splunk) [reply exporter/exportertest/contract_checker.go]: Done with `assert.EventuallyWithT`, so I didn't need to rewrite functions to return bool

File: exporter/exportertest/contract_checker.go [hunk 14]
```diff
@@ -0,0 +1,271 @@
...
+func CreateOneTraceWithID(id UniqueIDAttrVal) ptrace.Traces {
+	data := ptrace.NewTraces()
+	data.ResourceSpans().AppendEmpty().ScopeSpans().AppendEmpty().Spans().AppendEmpty().Attributes().PutStr(
+		UniqueIDAttrName,
+		string(id),
+	)
+	return data
+}
+
+func CreateOneMetricWithID(id UniqueIDAttrVal) pmetric.Metrics {
```
Comment (dmitryax) [inline exporter/exportertest/contract_checker.go:267]: Should not be exposed. Limiting the API should the primary goal. Otherwise, it'll be hard to maintain and change after 1.0 release

File: exporter/exportertest/mock_receiver.go [hunk 15]
```diff
@@ -0,0 +1,237 @@
...
+	if r < third {
+		return consumererror.NewPermanent(errPermanent)
+	}
+	if r < 2*third {
+		return errNonPermanent
+	}
+	return nil
+}
+
+type BaseMockConsumer interface {
```
Comment (dmitryax) [inline exporter/exportertest/mock_receiver.go:68]: The interface is not being used. I think we should use it instead of exposing the struct `MockConsumer`. The interface will be `MockConsumer`, and struct will be unexported `mockConsumer`

File: exporter/exportertest/mock_receiver.go [hunk 16]
```diff
@@ -0,0 +1,237 @@
...
+type MockConsumer struct {
+	reqCounter          RequestCounter
+	mux                 sync.Mutex
+	exportErrorFunction func() error
+	ReceivedTraces      []ptrace.Traces
+	ReceivedMetrics     []pmetric.Metrics
+	ReceivedLogs        []plog.Logs
+}
+
+type MockConsumerFactory func() MockConsumer
```
Comment (dmitryax) [inline exporter/exportertest/mock_receiver.go:84]: This looks like not needed

File: exporter/exportertest/mock_receiver.go [hunk 17]
```diff
@@ -0,0 +1,237 @@
...
+
+type DecisionFunc func() error
+
+type MockReceiver interface {
+	Start() error
+	Stop() error
+	RequestCounter() RequestCounter
+}
+
+type MockReceiverFactory func(DecisionFunc) MockReceiver
```
Comment (dmitryax) [inline exporter/exportertest/mock_receiver.go:33]: I think we can avoid exposing `DecisionFunc` by changing this signature to 
```suggestion
type MockReceiverFactory func(MockConsumer) MockReceiver
```

CreateDefaultConsumer will not be needed after that

**Review by dmitryax — COMMENTED** [review]
Also, please rebase

File: exporter/otlpexporter/consume_contract_test.go (hunk 8 above)
Comment (dmitryax) [reply exporter/otlpexporter/consume_contract_test.go]: Keeping it hardcoded is ok

File: exporter/otlpexporter/consume_contract_test.go [hunk 18]
```diff
@@ -0,0 +1,100 @@
...
+	}
+}
+
+func testConfig() component.Config {
+	return &Config{
+		TimeoutSettings: exporterhelper.TimeoutSettings{},
+		QueueSettings:   exporterhelper.QueueSettings{Enabled: false},
+		RetrySettings:   newTestRetrySettings(),
+		GRPCClientSettings: configgrpc.GRPCClientSettings{
+			Endpoint: confignet.NetAddr{Endpoint: "0.0.0.0:4317", Transport: "tcp"}.Endpoint,
```
[the remark this answers is not in the document]
Comment (omrozowicz-splunk) [reply exporter/otlpexporter/consume_contract_test.go]: done

File: exporter/exportertest/contract_checker.go (hunk 14 above)
Comment (omrozowicz-splunk) [reply exporter/exportertest/contract_checker.go]: changes, I think nothing left is uppercase unnecessarily

File: exporter/exportertest/mock_receiver.go (hunk 15 above)
Comment (omrozowicz-splunk) [reply exporter/exportertest/mock_receiver.go]: I expanded it with all the methods we need:

```
type MockConsumer interface {
	ConsumeLogs(_ context.Context, ld plog.Logs) error
	ConsumeMetrics(_ context.Context, md pmetric.Metrics) error
	ConsumeTraces(_ context.Context, td ptrace.Traces) error
	Capabilities() consumer.Capabilities
	Clear()
	RequestCounter() RequestCounter
}
```

and I use it in `mockOtlpReceiver` like this:

```
type mockOtlpReceiver struct {
	mockConsumer    exportertest.MockConsumer
	traceReceiver   receiver.Traces
	metricsReceiver receiver.Metrics
	logReceiver     receiver.Logs
}
```

File: exporter/exportertest/mock_receiver.go (hunk 16 above)
Comment (omrozowicz-splunk) [reply exporter/exportertest/mock_receiver.go]: removed

File: exporter/exportertest/mock_receiver.go (hunk 17 above)
Comment (omrozowicz-splunk) [reply exporter/exportertest/mock_receiver.go]: I changed it this way, but I think we still need a way to create MockConsumer to pass decision function based on test suite, so I created `NewMockConsumer`:

```
func NewMockConsumer(decisionFunc func() error) MockConsumer {
	return &mockConsumer{
		reqCounter:          newRequestCounter(),
		mux:                 sync.Mutex{},
		exportErrorFunction: decisionFunc,
		ReceivedTraces:      nil,
		ReceivedMetrics:     nil,
		ReceivedLogs:        nil,
	}
}
```

Didn't figure out how to create `MockConsumer` without an exported function. Let me know if I misunderstood 🙏

**Comment by omrozowicz-splunk** [comment]
> Also, please rebase

done

File: exporter/otlpexporter/mock_receiver.go [hunk 19]
```diff
@@ -0,0 +1,64 @@
...
+}
+
+func (bor *mockOtlpReceiver) Stop() error {
+	bor.mockConsumer.Clear()
+	err := bor.traceReceiver.Shutdown(context.Background())
+	err = errors.Join(err, bor.metricsReceiver.Shutdown(context.Background()))
+	return errors.Join(err, bor.logReceiver.Shutdown(context.Background()))
+}
+
+func (bor *mockOtlpReceiver) RequestCounter() exportertest.RequestCounter {
```
Comment (dmitryax) [inline exporter/otlpexporter/mock_receiver.go:62]: The MockConsumer has all the information about received/rejected data, right? So we don't need to require implementation of this method and don't need to expose RequestCounter in the `exportertest`

File: exporter/exportertest/mock_receiver.go [hunk 20]
```diff
@@ -0,0 +1,235 @@
...
+	}
+	return nil
+}
+
+type MockConsumer interface {
+	ConsumeLogs(_ context.Context, ld plog.Logs) error
+	ConsumeMetrics(_ context.Context, md pmetric.Metrics) error
+	ConsumeTraces(_ context.Context, td ptrace.Traces) error
+	Capabilities() consumer.Capabilities
+	Clear()
```
Comment (dmitryax) [inline exporter/exportertest/mock_receiver.go:71]: Clear() is also not needed. It can be called in this package right after MockReceiver.Stop() if needed, instead of having it called by the API user from the Stop() method.

File: exporter/otlpexporter/mock_receiver.go (hunk 19 above)
Comment (omrozowicz-splunk) [reply exporter/otlpexporter/mock_receiver.go:62]: To clarify - do you mean we don't need a request counter at all? This is the only part that contain information about nonpermanent and permanent errors. According to the implementation I have now we need it exported to check if the test finished successfully:

```
reqCounter := mockReceiver.RequestCounter()
// The overall number of requests sent by exporter
fmt.Printf("Number of export tries: %d\n", reqCounter.total)
// Successfully delivered items
fmt.Printf("Total items received successfully: %d\n", reqCounter.success)
// Number of errors that happened
fmt.Printf("Number of permanent errors: %d\n", reqCounter.error.permanent)
fmt.Printf("Number of non-permanent errors: %d\n", reqCounter.error.nonpermanent)

assert.EventuallyWithT(params.T, func(c *assert.CollectT) {
	checkIfTestPassed(params.T, params.NumberOfTestElements, reqCounter)
}, 2*time.Second, 100*time.Millisecond)
```

MockConsumer indeed has all the information, but it's not exposed at all in `exportertest/contract_checker.go`, that's why I use it through `MockReceiver`. Do you suggest moving this check logic somewhere else?

Comment (dmitryax) [reply exporter/otlpexporter/mock_receiver.go:62]: > MockConsumer indeed has all the information, but it's not exposed at all in exportertest/contract_checker.go, that's why I use it through MockReceiver.

It's available in the `reqCounter` and it's not required to be exposed because it's being used in the same package - the checker, not on the API consumer side. You can convert the `MockConsumer` interface into the `mockConsumer` struct to get access to all the private field. But that maybe too hacky. I think we can expose the struct instead of the interface to make it easier: so remove `MockReceiver` interface and expose the `mockConsumer` struct -> `MockConsumer`. Sorry, looks like I'm asking to undo what I asked before, but we need to get the the point where the API is clean :)

Comment (omrozowicz-splunk) [reply exporter/otlpexporter/mock_receiver.go]: Ok, so I deleted MockConsumer interface completely, exposed MockConsumer struct instead, but I didn't delete `MockReceiver` interface as I thought we'd reuse it for otlp http exporter. Now I use request counter like this: `reqCounter := mockConsumer.reqCounter` and the MockReceiver interface is:

```
type MockReceiver interface {
	Start() error
	Stop() error
}
```

File: exporter/exportertest/mock_receiver.go (hunk 20 above)
Comment (omrozowicz-splunk) [reply exporter/exportertest/mock_receiver.go]: Now I use `mockConsumer.clear()` directly after 
```
err = mockReceiver.Stop()
require.NoError(params.T, err)
```

File: receiver/otlpreceiver/go.mod [hunk 21]
```diff
@@ -25,7 +25,6 @@ require (
 )
 
 require (
-	cloud.google.com/go/compute/metadata v0.2.4-0.20230617002413-005d2dfb6b68 // indirect
```
Comment (omrozowicz-splunk) [inline receiver/otlpreceiver/go.mod:28]: Not sure what is wrong, but every time I run `make gotidy` I have this error:
```
go.opentelemetry.io/collector/exporter/otlpexporter imports
        go.opentelemetry.io/collector/config/configgrpc imports
        go.opentelemetry.io/contrib/instrumentation/google.golang.org/grpc/otelgrpc tested by
        go.opentelemetry.io/contrib/instrumentation/google.golang.org/grpc/otelgrpc.test imports
        google.golang.org/grpc/interop imports
        golang.org/x/oauth2/google imports
        cloud.google.com/go/compute/metadata: ambiguous import: found package cloud.google.com/go/compute/metadata in multiple modules:
        cloud.google.com/go v0.26.0 (/Users/omrozowicz/go/pkg/mod/cloud.google.com/go@v0.26.0/compute/metadata)
        cloud.google.com/go/compute/metadata v0.2.4-0.20230617002413-005d2dfb6b68 (/Users/omrozowicz/go/pkg/mod/cloud.google.com/go/compute/metadata@v0.2.4-0.20230617002413-005d2dfb6b68)
make: *** [tidy] Error 1
```
For both otlp receiver and otlp exporter. When I bring back the changes from `main`, `make gotidy` deletes this automatically.

I inserted this line to make CI green but I am sure it's not a way to go:
```
replace cloud.google.com/go => cloud.google.com/go v0.110.2
```

File: exporter/exportertest/mock_receiver.go [hunk 22]
```diff
@@ -0,0 +1,221 @@
...
+	}
+}
+
+func newRequestCounter() requestCounter {
+	return requestCounter{
+		success: 0,
+		error:   newErrorCounter(),
+		total:   0,
+	}
+}
```
Comment (dmitryax) [inline exporter/exportertest/mock_receiver.go:109]: Please move it after the `MockConsumer` methods

File: receiver/otlpreceiver/go.mod (hunk 21 above)
Comment (dmitryax) [reply receiver/otlpreceiver/go.mod:28]: `replace cloud.google.com/go => cloud.google.com/go v0.110.2` is good. We have it all around the codebase

**Review by dmitryax — COMMENTED** [review]
I think we're almost done :) Thank you, @omrozowicz-splunk

File: exporter/otlpexporter/mock_receiver.go (hunk 19 above)
Comment (dmitryax) [reply exporter/otlpexporter/mock_receiver.go]: Sounds great!

File: receiver/otlpreceiver/go.mod [hunk 23]
```diff
@@ -130,6 +129,8 @@ replace go.opentelemetry.io/collector/extension/zpagesextension => ../../extensi
 
 replace go.opentelemetry.io/collector/consumer => ../../consumer
 
+replace cloud.google.com/go => cloud.google.com/go v0.110.2
```
[the remark this answers is not in the document]
Comment (omrozowicz-splunk) [reply receiver/otlpreceiver/go.mod:133]: done

File: exporter/exportertest/mock_receiver.go [hunk 24]
```diff
@@ -0,0 +1,221 @@
...
+	return nil
+}
+
+type MockConsumer struct {
+	reqCounter          requestCounter
+	mux                 sync.Mutex
+	exportErrorFunction func() error
+	ReceivedTraces      []ptrace.Traces
+	ReceivedMetrics     []pmetric.Metrics
+	ReceivedLogs        []plog.Logs
```
[the remark this answers is not in the document]
Comment (omrozowicz-splunk) [reply exporter/exportertest/mock_receiver.go]: done

File: exporter/exportertest/mock_receiver.go (hunk 22 above)
Comment (omrozowicz-splunk) [reply exporter/exportertest/mock_receiver.go]: done

File: exporter/exportertest/mock_receiver.go [hunk 25]
```diff
@@ -0,0 +1,221 @@
...
+type MockConsumer struct {
+	reqCounter          requestCounter
+	mux                 sync.Mutex
+	exportErrorFunction func() error
+	receivedTraces      []ptrace.Traces
+	receivedMetrics     []pmetric.Metrics
+	receivedLogs        []plog.Logs
+}
+
+func NewMockConsumer(decisionFunc func() error) MockConsumer {
```
Comment (dmitryax) [inline exporter/exportertest/mock_receiver.go:74]: This not needed to be exported I believe.

File: exporter/exportertest/mock_receiver.go [hunk 26]
```diff
@@ -0,0 +1,221 @@
...
+	"go.opentelemetry.io/collector/pdata/ptrace"
+)
+
+var errNonPermanent = status.Error(codes.DeadlineExceeded, "non Permanent error")
+var errPermanent = status.Error(codes.Internal, "Permanent error")
+
+type MockReceiver interface {
+	Start() error
+	Stop() error
+}
```
Comment (dmitryax) [inline exporter/exportertest/mock_receiver.go:28]: We can avoid introducing this interface and use `component.Component` instead.

**Review by dmitryax — COMMENTED** [review]
@omrozowicz-splunk couple last comments. Please rebase and I think we can merge it. Thank you!

File: exporter/exportertest/mock_receiver.go (hunk 25 above)
Comment (omrozowicz-splunk) [reply exporter/exportertest/mock_receiver.go]: done

File: exporter/exportertest/mock_receiver.go (hunk 26 above)
Comment (omrozowicz-splunk) [reply exporter/exportertest/mock_receiver.go]: done

**Review by dmitryax — APPROVED** [review]
LGTM. Thanks @omrozowicz-splunk!

**Comment by dmitryax** [comment]
@omrozowicz-splunk one last thing. Can you please add `contract_checker_test.go` with some basic test coverage. Looks like usage from another module isn't counted and it contract_checker_test.go the project's coverage below the threshold. Similar to `receivertest/contract_checker_test.go`

**Comment by omrozowicz-splunk** [comment]
> @omrozowicz-splunk one last thing. Can you please add `contract_checker_test.go` with some basic test coverage. Looks like usage from another module isn't counted and it contract_checker_test.go the project's coverage below the threshold. Similar to `receivertest/contract_checker_test.go` which should showcase how to use the helper for other exporters

Okaay, I can try to do that, but that will prolong this task even more - it requires a fake receiver and fake exporter and making the fake exporter send data to that fake receiver. Not sure if that will be super helpful, probably looking at the otlp example will be easier to understand. 
But I'll take a look, maybe it is easier than I imagine now 😅

**Comment by dmitryax** [comment]
Unfortunately, if we merge this without the tests, the collector's build will marked as failing because the project's coverage will be below the threshold

**Comment by omrozowicz-splunk** [comment]
Hey @dmitryax, I added `contract_checker_test.go`, it's probably not the most elegant, but works and will increase the coverage. I needed to change `requestCounter` in `MockConsumer` struct to `*requestCounter` to be able to control it from exporter and then passing over to receiver. Please take a look 😄

**Comment by dmitryax** [comment]
Thanks @omrozowicz-splunk! I added another commit to reduce the exposed API more and to simplify some tests. Let me know if something doesn’t look right for you.
